### PR TITLE
Extended duration symbols in ABC chords (marian-schultz)

### PR DESCRIPTION
**Pull request by marian-schulz**
Extension for the recognition of chords in ABCFormat according to http://abcnotation.com/wiki/abc:standard:v2.1#chords_and_unisons.

* For chords an outer length modifier is supported
Example:
[ceg]2 == [c2e2g2]
[c3e3g3]2 == [ceg]6 == [c6e6g6]

* The length of the chord is determined by its first note
Example:
[c2e4g8] == [c2e2g2]
* Empty chords are discarded
* Additional unittests for variations of abc Chords
* Fix: Missing chord symbols  for chords
* Additional unittests for chord symbols on note & chords

### CODE REVIEW

**Comment by mscuthbert** [comment]
coveralls had a problem -- closing and reopening to get a fresh test.

**Comment by mscuthbert** [comment]
Thanks -- look at the lint errors, etc. -- but it seems to be on the right track!

**Comment by marian-schulz** [comment]
I have fixed the PEP8 violations and cleaned up my commit history.
The tests are all green now.

I apologize for not doing the code inspection before.

**Comment by marian-schulz** [comment]
The chord symbols before a chord were missing. I have also fixed this and added an additional unittest.

File: music21/abcFormat/testFiles.py
Comment (mscuthbert) [inline music21/abcFormat/testFiles.py:834]: see below about elements[1]

File: music21/abcFormat/testFiles.py
Comment (mscuthbert) [inline music21/abcFormat/testFiles.py:857]: can `.elements[1]` be changed to whatever it is we're actually getting, like `.parts[0]` etc. -- otherwise this test is fragile if other metadata elements or something else like a spanner is later a default part of abcToStreamScore.

File: music21/abcFormat/testFiles.py
Comment (mscuthbert) [inline music21/abcFormat/testFiles.py:846]: Add some tests with sharps, accidentals, etc. to make sure that c#/2 still works.  Maybe a separate test so it doesn't break the pitch_name in 'CEG' test below.  Just always paranoid about that.

File: music21/abcFormat/testFiles.py
Comment (mscuthbert) [inline music21/abcFormat/testFiles.py:862]: great!

**Review by mscuthbert — CHANGES_REQUESTED** [review]
Looks great!  The `.elements[1]` call is all that needs to be changed, to make sure that it's explicitly getting the type of element that it should be.

(btw -- `s.elements[1]` is almost always better as `e[1]` anyhow -- .elements[1] is slower and doesn't setActiveSite).

File: music21/abcFormat/testFiles.py
Comment (marian-schulz) [reply music21/abcFormat/testFiles.py]: Fixed this in all cases. I use the getElementbyClass of operator now.

File: music21/abcFormat/testFiles.py
Comment (marian-schulz) [reply music21/abcFormat/testFiles.py]: I agree.

**Comment by marian-schulz** [comment]
Optimze the Testcases.
+ Use the ElementsByClass operator instead of an index access to .elements[x].
+ Add sharps, flats, natural and ocatve modificators to notes.

Broken rythm
+ Fixes broken rythm for Chords (i guess i broke that)
+ Add A Testcase to cover brocken rythm between chords and notes

File: music21/abcFormat/__init__.py
Comment (jacobtylerwalls) [inline music21/abcFormat/__init__.py:1655]: Chords

File: music21/abcFormat/testFiles.py
Comment (jacobtylerwalls) [inline music21/abcFormat/testFiles.py:941]: Better to leave this commented so it doesn't launch a file.

**Review by jacobtylerwalls — COMMENTED** [review]
Happened to stroll by and noticed a couple small typos or details.

File: music21/abcFormat/testFiles.py
Comment (mscuthbert) [inline music21/abcFormat/testFiles.py:834]: this is returning a Stream iterator -- you probably want a specific part, like:

```
part = s.getElementsByClass(stream.Part)[0]
```

which is abbreviated as:

```
part = s.parts[0]
```

File: music21/abcFormat/testFiles.py
Comment (mscuthbert) [inline music21/abcFormat/testFiles.py]: Similarly here.  I'll bet that `e` is never being set.  Temporarily put a print statement into the loop and you'll see it never runs.

File: music21/abcFormat/__init__.py
Comment (mscuthbert) [inline music21/abcFormat/__init__.py:1674]: misalignment of hanging indent (one space more)

File: music21/abcFormat/testFiles.py
Comment (marian-schulz) [reply music21/abcFormat/testFiles.py:834]: Thanks,
I fixed this and  notice an  additional Bug, i guess made by the translater. (I still have to research this)

for abc_chord in ['[]', '[z]']:
    ah = af.readstr(abc_dl + '[]')
    s = translate.abcToStreamScore(ah)
    chord = list(s.parts[0].notes)
    print(abc_chord, chord, ah.tokens)

[] [<music21.note.Note C>] [<music21.abcFormat.ABCMetadata 'L:1/8'>, <music21.abcFormat.ABCChord '[]'>]
[z] [<music21.note.Note C>] [<music21.abcFormat.ABCMetadata 'L:1/8'>, <music21.abcFormat.ABCChord '[]'>]

There is no Chord at all but a note C in the part. The C is not present in abc code nor token list at all.

File: music21/abcFormat/__init__.py
Comment (marian-schulz) [reply music21/abcFormat/__init__.py]: add this extra space and removed the previous line.

File: music21/abcFormat/testFiles.py
Comment (marian-schulz) [reply music21/abcFormat/testFiles.py]: I wonder while the stream iterator implements .getElementsByClass at all ?
However .parts[0] looks better.
fixed in the last commit.

Comment (mscuthbert) [reply music21/abcFormat/testFiles.py]: It totally does.  StreamIterators are designed to be chained, so most explicit cases would be:

```
s.flat.iter.getElementsByOffset(10.0, 20.0).getElementsByClass('Note').getElementsByGroup('cue').addFilter(lambda n: n.pitch.alter == 1)
```

As of v.6., each of these "getElementsBy..." returns a new StreamIterator that adds a new filter on the notes.  So `s.getElementsByClass(stream.Part).getElementsByClass(chord.Chord)` would always return no elements, because first it has to be of class Part, and then of class Chord, and nothing will ever be like that.

It might be that adding a second filter of the same type should raise an exception, but there are a few cases where you might want this, such as:

`s.getElementsByClass(spanner.Spanner).getElementsByClass(expressions.Tremolo)`

which only gets Tremolos that are Spanners. 

Maybe that doesn't work, but there are a few cases like that.

**Comment by mscuthbert** [comment]
looks really good -- one little change -- capital letter variable names are only used for classes.  Will merge if everything passes.  Thank you for the great contributions!

File: music21/abcFormat/testFiles.py
Comment (jacobtylerwalls) [inline music21/abcFormat/testFiles.py:887]: rhythm

File: music21/abcFormat/testFiles.py
Comment (jacobtylerwalls) [inline music21/abcFormat/testFiles.py:895]: same

File: music21/abcFormat/testFiles.py
Comment (jacobtylerwalls) [inline music21/abcFormat/testFiles.py:939]: here and above

**Review by jacobtylerwalls — COMMENTED** [review]
Noticed a few instances where "rhythm" is misspelled.

**Comment by mscuthbert** [comment]
Please don't force-push when making Pull Requests -- it makes it impossible to review what has changed.  Your open-source maintainers will appreciate it.

**Comment by mscuthbert** [comment]
Congrats! Merged

**Pull request merged**
